Source/WebCore/platform/graphics/gstreamer/ImageDecoderGStreamer.cpp

```diff
@@ -318,131 +297,63 @@ void ImageDecoderGStreamer::notifySample(GRefPtr<GstSample>&& sample)
-                    decoder.m_messageCondition.wait(decoder.m_messageLock);
-            }
-        }
-        gst_message_unref(message);
-        return GST_BUS_DROP;
-    }, this, nullptr);
+    });
 
-    GstElement* source = makeGStreamerElement("giostreamsrc", nullptr);
-    g_object_set(source, "stream", m_memoryStream.get(), nullptr);
+    auto caps = adoptGRef(gst_type_find_helper_for_buffer(GST_OBJECT_CAST(m_parserHarness->element()), buffer, nullptr));
+    GST_DEBUG_OBJECT(m_parserHarness->element(), "Caps typefind result: %" GST_PTR_FORMAT, caps.get());
+    if (!caps) {
+        GST_WARNING_OBJECT(m_parserHarness->element(), "Typefinding failed");
+        return;
+    }
 
-    m_decodebin = makeGStreamerElement("decodebin3", nullptr);
-    g_signal_connect_swapped(m_decodebin.get(), "pad-added", G_CALLBACK(decodebinPadAddedCallback), this);
+    m_parserHarness->start(WTFMove(caps));
+    m_parserHarness->pushBuffer(buffer);
 
-    gst_bin_add_many(GST_BIN_CAST(m_pipeline.get()), source, m_decodebin.get(), nullptr);
-    gst_element_link(source, m_decodebin.get());
-    gst_element_set_state(m_pipeline.get(), GST_STATE_PLAYING);
-}
+    if (!m_decoderHarness) {
+        GST_WARNING_OBJECT(m_parserHarness->element(), "Parsing failed");
+        return;
+    }
 
-void ImageDecoderGStreamer::InnerDecoder::run()
-{
-    m_runLoop.dispatch([this]() {
-        preparePipeline();
-    });
+    for (auto& stream : m_parserHarness->outputStreams()) {
+        while (auto event = stream->pullEvent())
+            m_decoderHarness->pushEvent(event.leakRef());
+    }
 
-    m_runLoop.run();
+    m_decoderHarness->flush();
 }
 
 EncodedDataStatus ImageDecoderGStreamer::InnerDecoder::encodedDataStatus() const
 {
-    GstState state;
-    gst_element_get_state(m_pipeline.get(), &state, nullptr, 0);
-    if (state >= GST_STATE_READY)
+    if (m_parserHarness->inputCaps())
         return EncodedDataStatus::TypeAvailable;
     return EncodedDataStatus::Unknown;
 }
 
+void ImageDecoderGStreamer::storeDecodedSample(GRefPtr<GstSample>&& sample)
+{
+    auto presentationSize = getVideoResolutionFromCaps(gst_sample_get_caps(sample.get()));
+    if (presentationSize && !presentationSize->isEmpty() && (!m_size || m_size != roundedIntSize(*presentationSize)))
+        m_size = roundedIntSize(*presentationSize);
+    m_sampleData.addSample(ImageDecoderGStreamerSample::create(WTFMove(sample), *m_size));
+}
+
 void ImageDecoderGStreamer::pushEncodedData(const FragmentedSharedBuffer& buffer)
 {
-    auto contiguousBuffer = buffer.makeContiguous();
+    auto data = buffer.makeContiguous()->extractData();
     m_eos = false;
-    auto thread = Thread::create("ImageDecoderGStreamer", [this, data = contiguousBuffer->data(), size = buffer.size()] {
-        m_innerDecoder = ImageDecoderGStreamer::InnerDecoder::create(*this, data, size);
+    auto thread = Thread::create("ImageDecoderGStreamer", [this, buffer = Span<const uint8_t> { data.data(), data.size() }] {
```

Comment:
I guess you're assuming that this this lambda you pass to `Thread::create` is going to stay alive, probably with the `thread`, at least until it is run because otherwise you'd need to not use a `Span` and take ownership of that data into to the lambda.